azavea/franklin
# Support query extension

## Overview

This PR adds (will add) support for the [query extension](https://github.com/radiantearth/stac-api-spec/tree/master/extensions/query), which allows querying STAC items by arbitrary fields in their `properties`.

### Checklist


### Notes

It's not really clear what should happen with poorly typed queries. If there are two items with incompatible field types (e.g., one has `"foo": 3` and the other has `"foo": "bar"`, I don't think it will be possible to filter on `foo` because postgres will freak out about the invalid comparison for one of them. One option is just to make everything a `String` at filter time I guess, but it feels bad. I'm hopeful that collection filters will prevent this from happening too frequently but it's tough to know that for certain.

~Also the valid types for queries aren't super well-defined. Can I do `"eq": ["some", "array"]`? I don't know! With the preference against nesting I'm assuming that the query extension is primarily for fields that have string or numeric values, but I don't know about arrays.~ It turns out I can dodge this problem entirely with `jsonb` equality checks. It should also fix the inconsistent postgres types issue I think.

### Testing Instructions

- import the berlin data: `bloop run application -- import --catalog-root s3://rasterfoundry-development-data-us-east-1/berlin-catalog/catalog.json --db-host localhost`
- save this json to `filters.json`:

```json
{
    "bbox": null,
    "datetime": null,
    "intersects": null,
    "collections": ["berlin"],
    "query": {
	"datetime": {
	    "eq": "2018-03-03T10:20:19Z"
	}
    }
}
```

- start up the server
- `cat filters.json | http :9090/search` -- you should get the `image` stac item back
- change the `eq` to a `neq` -- you should get no results back
- more exciting: change `datetime` to `label:type` and the datetime string to `"raster"` -- you should get items back with `label:type` equal to `"raster"`

Closes #36
